internal/session/session.go

```diff
@@ -48,144 +71,151 @@ type Body struct {
+	ticker := time.NewTicker(1)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-s.ctx.Done():
+			log.Logger.Debug("session writer: closing writer")
+			s.writerClosedCh <- true
+			log.Logger.Debug("session writer: closed writer")
+			return
+
+		case <-ticker.C:
+			ticker.Reset(s.pipeInterval)
+		}
 
-			reader, writer := io.Pipe()
-			req, err := http.NewRequest("POST", serverUrl, reader)
-			if err != nil {
-				log.Logger.Errorf("session writer: error creating request: %v, retrying in 3s...", err)
-				time.Sleep(3 * time.Second)
-				continue
-			}
-			req.Header.Set("machine_id", s.machineID)
-			req.Header.Set("session_type", "write")
-
-			log.Logger.Debugf("session writer: created")
-			goroutineCloseCh := make(chan any)
-			go func() {
-				for {
-					select {
-					case <-s.writerCloseCh:
-						writer.Close()
-					case <-goroutineCloseCh:
-						return
-					case body := <-s.writer:
-						bytes, err := json.Marshal(body)
-						if err != nil {
-							log.Logger.Errorf("session writer: failed to marshal body: %v", err)
-							continue
-						}
-						if _, err := writer.Write(bytes); err != nil {
-							log.Logger.Errorf("session writer: failed to write to pipe: %v", err)
-							continue
-						}
+		reader, writer := io.Pipe()
+		goroutineCloseCh := make(chan any)
+		go func() {
+			log.Logger.Debug("session writer created")
+			for {
+				select {
+				case <-s.writerCloseCh:
+					log.Logger.Debug("session writer closed")
+					writer.Close()
+					return
+				case <-goroutineCloseCh:
+					return
+				case body := <-s.writer:
+					bytes, err := json.Marshal(body)
+					if err != nil {
+						log.Logger.Errorf("session writer: failed to marshal body: %v", err)
+						continue
+					}
+					if _, err := writer.Write(bytes); err != nil {
+						log.Logger.Errorf("session writer: failed to write to pipe: %v", err)
+						continue
 					}
 				}
-			}()
-
-			client := &http.Client{}
-			resp, err := client.Do(req)
-			if err != nil {
-				log.Logger.Errorf("session writer: error making request: %v, retrying", err)
-				time.Sleep(3 * time.Second)
-				close(goroutineCloseCh)
-				continue
 			}
-			log.Logger.Debugf("session writer: unexpected closed, resp: %v %v, reconnecting...", resp.Status, resp.StatusCode)
+		}()
+
+		req, err := http.NewRequestWithContext(s.ctx, "POST", s.endpoint, reader)
+		if err != nil {
+			log.Logger.Errorf("session writer: error creating request: %v, retrying in 3s...", err)
+			close(goroutineCloseCh)
+			continue
+		}
+		req.Header.Set("machine_id", s.machineID)
+		req.Header.Set("session_type", "write")
+
+		client := &http.Client{}
+		resp, err := client.Do(req)
+		if err != nil {
+			log.Logger.Errorf("session writer: error making request: %v, retrying", err)
 			close(goroutineCloseCh)
-			time.Sleep(3 * time.Second)
+			continue
 		}
-	}()
+
+		log.Logger.Debugf("session writer: unexpected closed, resp: %v %v, reconnecting...", resp.Status, resp.StatusCode)
+		close(goroutineCloseCh)
```

Comment:
@cardyok  We didn't `close(goroutineCloseCh)`, now we do. Please check.

Reply:
this is intentionally left out to ensure that only one pair of read/write connection is there..

Reply:
@cardyok removed